### ISSUE TITLE
question about objectness head.

### ISSUE THREAD

**Issue by xiaomoguhzz** [orrzohar/PROB#53]
hi, you have done a excellent work!  Recently I'm trying to transplant your work to mine, and my research direction is open-vocabulary detection. In the open vocabulary detection, I encountered the problem that the model objectness confidence in the novel category was too low during inference(Classification is done by CLIP). This was caused by the fact that there was only base category annotation in the training phase. So I think the research on open-world detection can help me. But after I ported your ProbObjectnessHead to my code and the training was completed, the following prediction results appeared. I haven't found the reason yet. Can you help me? 
![微信图片_20240220214216](https://github.com/orrzohar/PROB/assets/31958916/6a1535ad-0cb5-49a2-bca6-874bff71f24d)
As you can see, the prediction results for ProbObjectnessHead look strange, with the maximum stuck at 0.3.

**Comment by orrzohar** [comment]
Hi @xiaomoguhzz,
Very interesting work! It would be very interesting to see if the ProbObjHead improves performance in OVD generalization. 

Yes this does not look OK. I want to sure of some things:
1.  This is how you covert the predicted objectness to object probability: `obj_prob = torch.exp(-self.temperature*pred_obj).unsqueeze(-1)`. Is that what you do?
2. While training, this is how you optimize, right? (i.e., you only apply the loss to predictions matched to known objects during training):
```
def loss_obj_likelihood(self, outputs, targets, indices, num_boxes):
    assert "pred_obj" in outputs
    idx = self._get_src_permutation_idx(indices)
    pred_obj = outputs["pred_obj"][idx]
    return  {'loss_obj_ll': torch.clamp(pred_obj, min=self.min_obj).sum()/ num_boxes}
```
3. As you are likely working with a radically different number of objects, objectness temperature will likely need to be optimized. Luckily, obj temperature does not affect training only inference, so once you have a trained model you can sweep this hyper parameter.
4. Note that in training, I weight the known and unknown/background logits differently: https://github.com/orrzohar/PROB/blob/28afbc1d7f5bdfb9a384ce09c870fcd829adc1b2/models/prob_deformable_detr.py#L343-L345
`self.empty_weight` weights down the unknown object predictions as PROB make 100 predictions per image, and on average there is ~10 objects per image, so we need to down-weight the known-background/unknown predictions 1:10.

I would be happy to help you debug as much as you need, 
Orr

**Comment by xiaomoguhzz** [comment]
> Hi @xiaomoguhzz, Very interesting work! It would be very interesting to see if the ProbObjHead improves performance in OVD generalization.
> 
> Yes this does not look OK. I want to sure of some things:
> 
> 1. This is how you covert the predicted objectness to object probability: `obj_prob = torch.exp(-self.temperature*pred_obj).unsqueeze(-1)`. Is that what you do?
> 2. While training, this is how you optimize, right? (i.e., you only apply the loss to predictions matched to known objects during training):
> 
> ```
> def loss_obj_likelihood(self, outputs, targets, indices, num_boxes):
>     assert "pred_obj" in outputs
>     idx = self._get_src_permutation_idx(indices)
>     pred_obj = outputs["pred_obj"][idx]
>     return  {'loss_obj_ll': torch.clamp(pred_obj, min=self.min_obj).sum()/ num_boxes}
> ```
> 
> 3. As you are likely working with a radically different number of objects, objectness temperature will likely need to be optimized. Luckily, obj temperature does not affect training only inference, so once you have a trained model you can sweep this hyper parameter.
> 4. Note that in training, I weight the known and unknown/background logits differently: https://github.com/orrzohar/PROB/blob/28afbc1d7f5bdfb9a384ce09c870fcd829adc1b2/models/prob_deformable_detr.py#L343-L345
>    
>    `self.empty_weight` weights down the unknown object predictions as PROB make 100 predictions per image, and on average there is ~10 objects per image, so we need to down-weight the known-background/unknown predictions 1:10.
> 
> I would be happy to help you debug as much as you need, Orr

hi @orrzohar , Thank you very much for taking the time to answer my questions. Here I give how I transplant your code to my project, and I hope you can help me see if there is any problem. 
1. first, I init it in my Deformable-DETR __init__ function，just as follows. In my code, class num is 1 since my model only gives the CLIP Classifier model a signal that how confident a predict box contain a object. This is also the reason why my model does not generalize enough, because there is only the base class annotation in the training stage, and the novel category will be regarded as the background when it appears in the training stage, resulting in very low confidence of novel category in the inference stage.
![微信图片_20240221100347](https://github.com/orrzohar/PROB/assets/31958916/df7b428f-be80-4e39-9948-4421726d5070)
2. similar to your code, I execute it after the transformer module. I kept it consistent with your code, which is using a shared ProbObjectnessHead in the 6-layer decoder. Hope I understood it correctly.
![微信图片_20240221100619](https://github.com/orrzohar/PROB/assets/31958916/74bbdde0-b617-4549-907a-407ee4b5693a)
3.  then I collect the final result and aux result:
![image](https://github.com/orrzohar/PROB/assets/31958916/bd7b087f-e252-40a6-8c1a-8b8c4945fcfb)
4. In fact, I have not run your code, but I see that your code uses the vanilla Hungarian matching algorithm, which is what I want to confirm. The result of ProbObjectnessHead does not participate in Hungarian matching.
5.  After matching, I use the matched prediction to compute "loss_obj_ll". Including final prediction results and auxiliary prediction results
![image](https://github.com/orrzohar/PROB/assets/31958916/bc8d0790-5818-4d3a-b623-601b25cc1c5a)
6. In fact, I noticed that you modified sigmoid_focal_loss yourself. If I understand correctly, you assigned the labels of the prediction results that were not matched by Hungarian matching to 80, which can be understood as background or unknown class. And adjust the sigmoid_focal_loss weight of category 80 to 0.1. I am confused about this part. In my code, the class num is 1, which represents the confidence of whether a prediction box contains an object. I have two options. One is to directly remove my original class head and multiply the result of ProbObjectnessHead with the CLIP classification result. Another option is to multiply the class result with the ProbObjectnessHead result and the CLIP classification result. My current code is written according to option 2.  I also tried to change the class to 2, which is consistent with your code, that is, category 1 is a known class, category 2 is an unknown class/background, and the weight of category 2 is changed to 0.1.  And the label of predictions that not be matched by the Hungarian matching algorithm is set to 2. But if I do this, I don’t know how to write the inference code, because originally, during inference, I only needed to multiply the class head score and the CLIP classification score, but now there are two class head scores.
![image](https://github.com/orrzohar/PROB/assets/31958916/265ab244-2b08-4f9f-ad7b-35e65596ad28)
7. During the inference phase, as you can see, I multiplied the prob results, the class results, and the CLIP classification results.
![image](https://github.com/orrzohar/PROB/assets/31958916/8cc7a936-6d48-4807-b554-a88993923201)
I have proposed quite a few questions, and I hope they don't offend you. I sincerely hope to receive your assistance.
xiaomoguhzz

**Comment by xiaomoguhzz** [comment]
And @orrzohar ,here is my training log screenshot，I compared the loss value of ProbObjectnessHead in the screenshot you gave. It seems that there is nothing abnormal about the loss value of ProbObjectnessHead in my code.
![image](https://github.com/orrzohar/PROB/assets/31958916/f56cceff-6ac3-47cf-a752-2250438d2745)

**Comment by xiaomoguhzz** [comment]
hi @orrzohar,  after debugging, I have found that when switching to eval() mode(i.e. ,Evaluate on the validation set ), the output of ProbObjectnessHead becomes non-discriminative.
![image](https://github.com/orrzohar/PROB/assets/31958916/3c4b602a-0c5e-4a1e-b7d0-0edf125bacdf)
however, when the model is in the training stage, I also print out the output of ProbObjectnessHead, which is very different from the inference stage, which looks quite strange. Looks like overfitting?
![image](https://github.com/orrzohar/PROB/assets/31958916/502db8f1-c62e-4bd4-8938-4c5dfde05e47)
here is my debug code location:
![image](https://github.com/orrzohar/PROB/assets/31958916/604e74cd-a7a0-496a-9a85-e95b9a57198d)

**Comment by orrzohar** [comment]
Hi @xiaomoguhzz,

I apologize for disappearing; I had a tough ECCV push.  Is this still relevant?

Also, feel free to contact me at orrzohar@stanford.edu. I would happily set up a 30-45-minute Zoom call to help you debug.

Best,
Orr
